### PR TITLE
Fix tf.keras.ops.numpy.logaddexp and logaddexp2 gradient at equal inputs

### LINKED ISSUE #127232: tf.keras.ops.numpy.logaddexp returns incorrect gradient at equal inputs
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.21.0, also reproduced on tf 2.22.0-dev20260904

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Mobile device

_No response_

### Python version

Python 3.13.5

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Current behavior?

`tf.keras.ops.numpy.logaddexp` returns an incorrect first derivative at equal inputs.

The reproducer evaluates `tf.keras.ops.numpy.logaddexp(t, [-4, 0, 5])` inside a weighted scalar expression at `t = -4.0`. In the first term, the two inputs to `logaddexp` are equal. Since `log(exp(x) + exp(y))` is smooth at equality, the partial derivative with respect to the first input should be `1/2` at that point. A stabilization branch should preserve this derivative.

The forward value is computed correctly as `-5.770614664699297`, but reverse-mode autodiff returns `1.0359107226361899` for the derivative with respect to `t`. The expected derivative is approximately `0.53591072263619`.

Expected behavior?

The first derivative of the weighted `tf.keras.ops.numpy.logaddexp` expression at `t = -4.0` should be close to `0.53591072263619`, not `1.0359107226361899`.

### Standalone code to reproduce the issue

```shell
import os
os.environ["CUDA_VISIBLE_DEVICES"] = ""
os.environ["OMP_NUM_THREADS"] = "1"
os.environ["TF_NUM_INTRAOP_THREADS"] = "1"
os.environ["TF_NUM_INTEROP_THREADS"] = "1"

import tensorflow as tf

def target(t):
    y = tf.keras.ops.numpy.logaddexp(
        t,
        tf.constant([-4, 0, 5], dtype=t.dtype),
    )
    return y[0] + 2 * y[1] - y[2] / 2

x = tf.constant(-4.0, dtype=tf.float64)

with tf.GradientTape() as tape:
    tape.watch(x)
    y = target(x)

actual = tape.gradient(y, x)
expected = 0.53591072263619

print("forward:", y.numpy())
print("actual:", actual.numpy())
print("expected:", expected)

if abs(float(actual) - expected) > 1e-9:
    print("BUG REPRODUCED: tf.keras.ops.numpy.logaddexp returns incorrect gradient at equal inputs")
else:
    print("not reproduced")
```

### Relevant log output

```shell
forward: -5.770614664699297
actual: 1.0359107226361899
expected: 0.53591072263619
BUG REPRODUCED: tf.keras.ops.numpy.logaddexp returns incorrect gradient at equal inputs
```

### PR DESCRIPTION
**Pull request by jdymitarai**
## Description

This PR fixes an automatic differentiation bug where `tf.keras.ops.numpy.logaddexp` and `logaddexp2` returned incorrect gradients when inputs were equal (fixes tensorflow/tensorflow#127232).

### Root Cause
In `keras/src/backend/tensorflow/numpy.py`, `logaddexp` and `logaddexp2` evaluated:
```python
amax = tf.maximum(x1, x2)
...
return tf.where(
    tf.math.is_nan(delta),
    x1 + x2,
    amax + tf.math.log1p(tf.math.exp(-tf.abs(delta))),
)
```
When $x_1 = x_2$:
1. `tf.maximum(x1, x2)` routes 100% of the upstream gradient to $x_1$ and 0% to $x_2$ because $x_1 \ge x_2$ is true.
2. The subgradient of `tf.abs(delta)` at $\delta = 0$ is $0$, which causes the second term `log1p(exp(-abs(delta)))` to produce zero gradient.
3. Consequently, reverse-mode autodiff at equal inputs collapsed to `(1.0, 0.0)` instead of the correct mathematical gradient `(0.5, 0.5)`.

### Solution
We reformulate the numerical stabilization using `safe_max = tf.stop_gradient(...)`, matching standard LogSumExp implementations:
```python
delta = x1 - x2
amax = tf.maximum(x1, x2)
finite_amax = tf.math.is_finite(amax)
finite_delta = tf.math.is_finite(delta)

safe_max = tf.stop_gradient(
    tf.where(finite_amax, amax, tf.zeros_like(amax))
)
# Mask inputs to safe_max when delta is non-finite so that the unselected
# LSE branch stays strictly finite and WhereGrad does not leak 0 * NaN.
x1_m = tf.where(finite_delta, x1, safe_max)
x2_m = tf.where(finite_delta, x2, safe_max)
lse = safe_max + tf.math.log(
    tf.math.exp(x1_m - safe_max) + tf.math.exp(x2_m - safe_max)
)
return tf.where(
    tf.math.is_nan(delta),
    x1 + x2,
    tf.where(finite_delta, lse, amax),
)
```
1. Because `safe_max` is stopped from backpropagation, reverse-mode autodiff evaluates:
$$\frac{\partial}{\partial x_1} \left( S + \ln(e^{x_1 - S} + e^{x_2 - S}) \right) = \frac{e^{x_1 - S}}{e^{x_1 - S} + e^{x_2 - S}} = \text{softmax}_1$$
At $x_1 = x_2$, both partial derivatives are strictly $0.5$.
2. When `delta` is infinite or non-finite (e.g. $(+\infty, 5.0)$), `tf.where(finite_delta, ...)` falls back to `amax`. Furthermore, `x1_m` and `x2_m` are masked with `safe_max` so the unselected `lse` branch evaluates strictly finite expressions (avoiding `inf / inf` divisions or `0 * NaN` gradient leakage).
3. The identical stabilization is applied to `logaddexp2`, preserving symbolic `log2` evaluation.

### Testing
- Added `test_logaddexp_gradients` and `test_logaddexp2_gradients` in `keras/src/ops/numpy_test.py` across TensorFlow, JAX, and PyTorch backends.
- Added test case validating the exact weighted expression from tensorflow/tensorflow#127232 (`0.53591072263619` with `atol=1e-7`).
- Added gradient assertions with infinite inputs (`+inf`, `-inf`) ensuring robust gradients across backends.

Fixes tensorflow/tensorflow#127232

## Contributor Agreement

- [x] This PR is linked to an issue that has been assigned to me (see `Fixes #xxx` above).
- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

### CODE REVIEW

File: keras/src/backend/tensorflow/numpy.py [hunk 1]
```diff
@@ -2133,10 +2133,16 @@ def logaddexp(x1, x2):
...
+        tf.where(tf.math.is_finite(amax), amax, tf.zeros_like(amax))
+    )
+    exp_1 = tf.math.exp(x1 - safe_max)
+    exp_2 = tf.math.exp(x2 - safe_max)
     return tf.where(
         tf.math.is_nan(delta),
         x1 + x2,
-        tf.maximum(x1, x2) + tf.math.log1p(tf.math.exp(-tf.abs(delta))),
+        safe_max + tf.math.log(exp_1 + exp_2),
     )
```
[the remark this answers is not in the document]
Comment (jdymitarai) [reply keras/src/backend/tensorflow/numpy.py:2152]: Applied the suggested 	f.where(tf.math.is_finite(delta), ...) bypass to eliminate NaN gradients at infinity and added corresponding gradient unit test assertions across backends. Thank you!

File: keras/src/backend/tensorflow/numpy.py [hunk 2]
```diff
@@ -2147,12 +2153,17 @@ def logaddexp2(x1, x2):
...
     log2 = tf.cast(tf.math.log(2.0), dtype)
+    exp_1 = tf.math.exp((x1 - safe_max) * log2)
+    exp_2 = tf.math.exp((x2 - safe_max) * log2)
     return tf.where(
         tf.math.is_nan(delta),
         x1 + x2,
-        tf.maximum(x1, x2)
-        + tf.math.log1p(tf.math.exp(-tf.abs(delta) * log2)) / log2,
+        safe_max + tf.math.log(exp_1 + exp_2) / log2,
     )
```
[the remark this answers is not in the document]
Comment (jdymitarai) [reply keras/src/backend/tensorflow/numpy.py:2181]: Applied the suggested 	f.where(tf.math.is_finite(delta), ...) bypass for logaddexp2 as well, maintaining symbolic log2 evaluation and adding infinite input gradient tests. Thank you!
